common/formatting/layout_optimizer.h

```diff
@@ -0,0 +1,445 @@
+// Copyright 2017-2021 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Improve code formatting by optimizing token partitions layout with
+// algorithm documented in https://research.google/pubs/pub44667/
+// (similar tool for R language: https://github.com/google/rfmt)
+
+#ifndef VERIBLE_VERILOG_FORMATTING_LAYOUT_OPTIMIZER_H_
+#define VERIBLE_VERILOG_FORMATTING_LAYOUT_OPTIMIZER_H_
+
+#include <deque>
+#include <vector>
+
+#include "common/formatting/basic_format_style.h"
+#include "common/formatting/token_partition_tree.h"
+#include "common/formatting/unwrapped_line.h"
+#include "common/util/vector_tree.h"
+
```

Comment:
It is probably worthwhile describing the idea of the layout optimizer here in a general header comment.

Also all classes below want a little documentation and how they interact with each other.
Think of looking at this code in half a year from now - do you still know what they're supposed to do ? Be kind to your future self :)